File: iosApp/iosApp/Modules/NFTsDashboard/NFTsDashboardPresenter.swift
// Created by web3d4v on 24/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib
import UniformTypeIdentifiers

enum NFTsDashboardPresenterEvent {
    case viewCollectionNFTs(collectionId: String)
    case viewNFT(identifier: String)
    case cancelError
    case sendError
}

protocol NFTsDashboardPresenter {
    func present(isPullDownToRefreh: Bool)
    func handle(_ event: NFTsDashboardPresenterEvent)
    func releaseResources()
}

final class DefaultNFTsDashboardPresenter {
    private weak var view: NFTsDashboardView!
    private let interactor: NFTsDashboardInteractor
    private let wireframe: NFTsDashboardWireframe
    private let nftsService: NFTsService
    
    private var latestNFTs: [NFTItem]?
    private var latestCollections: [NFTCollection]?
    private var error: Error?

    init(
        view: NFTsDashboardView,
        interactor: NFTsDashboardInteractor,
        wireframe: NFTsDashboardWireframe,
        nftsService: NFTsService
    ) {
        self.view = view
        self.interactor = interactor
        self.wireframe = wireframe
        self.nftsService = nftsService
        
        nftsService.addListener(self)
        interactor.addListener(self)
    }
    
    deinit {
        print("[DEBUG][Presenter] deinit \(String(describing: self))")
        interactor.removeListener(self)
    }
}

extension DefaultNFTsDashboardPresenter: NFTsDashboardPresenter {

    func present(isPullDownToRefreh: Bool) {
        if !isPullDownToRefreh {
            view.update(with: .loading)
        }
        
        interactor.fetchYourNFTs(
            isPullDownToRefreh: isPullDownToRefreh
        ) { [weak self] result in
            self?.handleYourNFTsResponse(with: result)
        }
    }

    func handle(_ event: NFTsDashboardPresenterEvent) {
        switch event {
        case let .viewCollectionNFTs(collectionId):
            wireframe.navigate(to: .viewCollectionNFTs(collectionId: collectionId))

        case let .viewNFT(identifier):
            guard let latestNFTs = latestNFTs else {
                return
            }
            let nftItem = latestNFTs.filter {
                        $0.identifier == identifier
                    }
                    .first
            guard let nftItem = nftItem else {
                return
            }
            wireframe.navigate(to: .viewNFT(nftItem: nftItem))

        case .cancelError:
            error = nil
            refreshView()

        case .sendError:
            UIPasteboard.general.setItems(
                [[UTType.utf8PlainText.identifier: errorMessage()]],
                options: [.expirationDate: Date().addingTimeInterval(30.0)]
            )
            wireframe.navigate(to: .sendError(msg: errorMessage()))
            error = nil
            refreshView()
        }
    }
    
    func releaseResources() {
        nftsService.removeListener(self)
    }
}

private extension DefaultNFTsDashboardPresenter {

    func handleYourNFTsResponse(with result: Result<[NFTItem], Error>) {
        switch result {
        case let .success(nfts):
            latestNFTs = nfts
            fetchNFTsCollections()
        case let .failure(err):
            error = err
            DispatchQueue.main.async { self.refreshView() }
        }
    }
    
    func fetchNFTsCollections() {
        interactor.fetchYourNFTsCollections { [weak self] result in
            guard let self = self else { return }
            self.handleCollectionsResponse(with: result)
        }
    }
    
    func handleCollectionsResponse(with result: Result<[NFTCollection], Error>) {
        switch result {
        case let .success(collections):
            latestCollections = collections
            refreshView()
        case let .failure(err):
            error = err
            DispatchQueue.main.async { self.refreshView() }
        }
    }
    
    func refreshView() {
        guard
            let latestNFTs = latestNFTs,
            let latestCollections = latestCollections
        else {
            if let error = error {
                let errorViewModel = EmbeddedErrorCollectionViewCell.ViewModel(
                    title: "Failed to load NFTs",
                    imageName: "",
                    message: "Unexpected error, please send error logs to developers",
                    button: "OK"
                )
                view.update(with: .error(errorViewModel))
            }
            return
        }

        let nfts: [NFTsDashboardViewModel.NFT] = latestNFTs.compactMap {
            NFTsDashboardViewModel.NFT(
                identifier: $0.identifier,
                image: $0.image
            )
        }
        
        let collections: [NFTsDashboardViewModel.Collection] = latestCollections.compactMap {
            NFTsDashboardViewModel.Collection(
                identifier: $0.identifier,
                coverImage: $0.coverImage,
                title: $0.title,
                author: $0.author
            )
        }

        view.update(
            with: .loaded(nfts: nfts, collections: collections)
        )
    }

    func errorMessage() -> String {
        let encodedDebugData = error
        return "Failed to load NFTs\n\nDebug data:\n\n\(encodedDebugData)"
    }
}

extension DefaultNFTsDashboardPresenter: NFTsServiceListener {
    
    func nftsChanged() {
        // Here we call to force a refresh in the view
        present(isPullDownToRefreh: true)
    }
}

extension DefaultNFTsDashboardPresenter: NFTsDashboardInteractorLister {
    
    func popToRootAndRefresh() {
        // Here we call to force a refresh in the view
        view.popToRootAndRefresh()
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/nftsDashboard/NFTsDashboardPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.nftsDashboard

import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3lib.utils.bgDispatcher
import com.sonsofcrypto.web3lib.utils.uiDispatcher
import com.sonsofcrypto.web3lib.utils.withUICxt
import com.sonsofcrypto.web3walletcore.common.viewModels.ErrorViewModel
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardPresenterEvent.ErrAction
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardPresenterEvent.Refresh
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardViewModel.Collection
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardViewModel.Error
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardViewModel.Loaded
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardViewModel.Loading
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardViewModel.NFT
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardWireframeDestination.SendError
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardWireframeDestination.ViewCollectionNFTs
import com.sonsofcrypto.web3walletcore.modules.nftsDashboard.NFTsDashboardWireframeDestination.ViewNFT
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

sealed class NFTsDashboardPresenterEvent {
    data class ViewCollectionNFTs(val idx: Int): NFTsDashboardPresenterEvent()
    data class ViewNFT(val idx: Int): NFTsDashboardPresenterEvent()
    data class ErrAction(val idx: Int): NFTsDashboardPresenterEvent()
    object Refresh: NFTsDashboardPresenterEvent()
}

interface NFTsDashboardPresenter {
    fun present()
    fun handle(event: NFTsDashboardPresenterEvent)
}

class DefaultNFTsDashboardPresenter(
    private val view: WeakRef<NFTsDashboardView>,
    private val wireframe: NFTsDashboardWireframe,
    private val interactor: NFTsDashboardInteractor,
): NFTsDashboardPresenter, NFTsDashboardInteractorLister  {
    private val bgScope = CoroutineScope(bgDispatcher)
    private val uiScope = CoroutineScope(uiDispatcher)
    private var isLoading: Boolean =  false
    private var err: Throwable? = null
    /** `errCache` needed due to potential race conditions, between user
     * handling error and new async events causing viewModel update */
    private var errCache: Throwable? = null

    init {
        interactor.add(this)
    }

    override fun present() {
        isLoading = true
        updateView()
        bgScope.launch {
            try {
                interactor.fetchYourNFTs()
                withUICxt { updateView() }
            } catch (e: Throwable) {
                err = e
                withUICxt { updateView() }
            }
        }
    }

    override fun handle(event: NFTsDashboardPresenterEvent) {
        when (event) {
            is NFTsDashboardPresenterEvent.ViewNFT -> {
                wireframe.navigate(ViewNFT(interactor.nfts()[event.idx]))
            }
            is NFTsDashboardPresenterEvent.ViewCollectionNFTs -> {
                val collId = interactor.collections()[event.idx].identifier
                wireframe.navigate(ViewCollectionNFTs(collId))
            }
            is Refresh -> present()
            is ErrAction -> {
                if (err != null && event.idx == 1) {
                    val e = err ?: errCache
                    val errMgs = Localized("nfts.dashboard.error.email.body", e)
                    wireframe.navigate(SendError(errMgs))
                }
                err = null
            }
        }
    }

    override fun networkChanged() {
        uiScope.launch { view.get()?.popToRootAndRefresh() }
    }

    override fun nftsChanged() { uiScope.launch { updateView() } }

    private fun updateView() {
        if (isLoading) {
            view.get()?.update(Loading)
            isLoading = false
            return
        }
        if (err != null) {
            println("[NFTsDashboardPresenter] error: $err")
            val errorViewModel = ErrorViewModel(
                Localized("error"),
                Localized("nfts.dashboard.error.message"),
                listOf(Localized("cancel"), Localized("sendLogs"))
            )
            view.get()?.update(Error(errorViewModel))
            errCache = err
            err = null
            return
        }
        val nftsViewModel = interactor.nfts().map {
            NFT(
                it.identifier,
                it.gatewayImageUrl,
                it.gatewayPreviewImageUrl,
                it.mimeType,
                it.fallbackText
            )
        }
        val collectionsViewModel = interactor.collections().map {
            Collection(it.identifier, it.coverImage, it.title, it.author)
        }
        view.get()?.update(Loaded(nftsViewModel, collectionsViewModel))
    }
}
